### PR TITLE
lxd: expose a documentation server to LXD

### PR DESCRIPTION
**Pull request by gabrielmougard**
JIRA task: https://warthogs.atlassian.net/jira/software/c/projects/LXD/boards/54?modal=detail&selectedIssue=LXD-359&assignee=63d9eb6128cddcc70770aff1

Spec: https://discourse.ubuntu.com/t/expose-the-full-lxd-documentation-website-as-part-of-a-lxd-deployment/37516

- [x] Expose documentation pages at `/documentation/`

Once this gets merged, we need to merge https://github.com/canonical/lxd-pkg-snap/pull/136 to make it work with snap

### CODE REVIEW

**Comment by monstermunchkin** [comment]
@gabrielmougard could you please fix the conflicts?

**Comment by gabrielmougard** [comment]
@simondeziel I realize something, about this `test/lint/embedded-full-doc-up-to-date.sh` test... now that `gendocs/html` contains pre-compressed assets, the md5 checksum will be inconsistent as gzip adds a timestamp and I'm not even sure that the compression algo is deterministic..

File: Makefile
Comment (simondeziel) [inline Makefile:128]: Could we have `PARALLEL="$(nproc)"` defined here and reused in later `find` calls? Subshells using backticks are deprecated BTW ;)

File: Makefile
Comment (simondeziel) [inline Makefile:131]: Could the HTML, JS, CSS and SVG minifiers pipe to `gzip -9` to write those assets only once?

File: Makefile
Comment (simondeziel) [inline Makefile:147]: If the HTML, CSS, JS and SVG files are compressed inline (see suggestion above), what else would that compress? If that final compression pass stays, I think it'd be safe to exclude `*.gz` filenames to avoid double compression.

File: lxd/gendocs/fileserver.go
Comment (simondeziel) [inline lxd/gendocs/fileserver.go:113]: If a client (say `wget`) doesn't advertise `gzip` support but tries to download an asset that is only available in pre-compressed form, will the server decompress the file on demand and serve it to the limited client? From my quick reading of the code fallback isn't implemented but I may have missed it.

File: Makefile
Comment (simondeziel) [inline Makefile:143]: AFAIK, PNG should not be `gzip` as that compression/optimization is done by `optipng` already.

File: lxd/gendocs/fileserver.go
Comment (gabrielmougard) [reply lxd/gendocs/fileserver.go:113]: This is indeed no implemented.  As of now, the code will stop there:
```go
func (f *fileHandler) findBestFile(w http.ResponseWriter, r *http.Request, name string) (http.File, os.FileInfo, error) {
   ae := r.Header.Get(acceptEncodingHeader)
      if ae == "" { // <--- a `wget` request without explicit encoding goes inside this case. 
         // Try to read the file without any encoding.
	 return f.openAndStat(name)
   }
   
   ...
}   
```

I think that `return f.openAndStat(name)` will give an error (as all the served files are gzipped, so if the file in the query is no the `.gz` one, this will fail), therefore ending in `http.NotFound()` error

Comment (simondeziel) [reply lxd/gendocs/fileserver.go:113]: OK then maybe pre-compression isn't something we should enable for now? Do you know how much size inflation we take by skipping the `gzip` pre-compression?

Comment (gabrielmougard) [reply lxd/gendocs/fileserver.go:113]: Without pre-compression, the additional size would be around **15MB**, instead of around **4.9MB**

Comment (gabrielmougard) [reply lxd/gendocs/fileserver.go:113]: And as we actually won't embed the website in the binary but rather in thee snap, this difference will not matter much right?

Comment (simondeziel) [reply lxd/gendocs/fileserver.go:131]: Considering the snap is `xz` compressed, I suspect those 15MB will compress even better than with `gzip`. Feels like a win-win because you then don't need to care about clients not supporting `gzip` but still have a moderate snap size inflation.

File: doc/api-extensions.md
Comment (tomponline) [inline doc/api-extensions.md:2289]: Did we have an API extension for `/ui`? If not then we dont need one for this, as its not really part of the API.

**Comment by tomponline** [comment]
@gabrielmougard when you're ready for another review please can you click the refresh review button on my name top right so that I get a notification and it'll appear on my todo list on https://github.com/pulls/review-requested

Thanks

**Comment by gabrielmougard** [comment]
@tomponline iit is successfully tested with the snap integration

File: doc/howto/access_documentation.md
Comment (ru-fu) [inline doc/howto/access_documentation.md:2]:
```suggestion
# How to access the local LXD documentation
```

Imo "web documentation" refers to the online documentation.

File: doc/howto/access_documentation.md
Comment (ru-fu) [inline doc/howto/access_documentation.md:6]:
```suggestion
Complete the following steps to access the local LXD documentation:
```

File: doc/howto/access_documentation.md
Comment (ru-fu) [inline doc/howto/access_documentation.md:11]:
```suggestion
1. Access the documentation in your browser by entering the server address followed by `/documentation` (for example, `https://192.0.2.10:8443/documentation/`).
```

File: po/ar.po
Comment (ru-fu) [inline po/ar.po:1]: Is there a UI string that needs reviewing?

**Review by ru-fu — CHANGES_REQUESTED** [review]

File: doc/howto/access_documentation.md
Comment (ru-fu) [inline doc/howto/access_documentation.md:13]:
```suggestion
1. Access the documentation in your browser by entering the server address followed by `/documentation/` (for example, `https://192.0.2.10:8443/documentation/`).
```
Sorry, missed that.

File: po/ar.po
Comment (ru-fu) [reply po/ar.po:1]: Is there?

File: lxd/api.go
Comment (tomponline) [inline lxd/api.go:82]: Can you see if the LXD_UI env var is documented, as we should probably document this in same place.

**Comment by tomponline** [comment]
Happy to merge this if @ru-fu is

File: doc/getting_started.md
Comment (ru-fu) [inline doc/getting_started.md:15]: These are ordered alphabetically so this needs to go before "Access the UI".

File: doc/howto/access_documentation.md
Comment (ru-fu) [inline doc/howto/access_documentation.md:5]:
```suggestion

Alternatively, you can access a local version of the LXD documentation that is embedded in the LXD snap.
```
Let's add a paragraph break here - looks better in the output.

**Review by ru-fu — COMMENTED** [review]
Still wondering if there are any UI strings I should have a look at (why are the translation files updated?).

**Comment by gabrielmougard** [comment]
> > Still wondering if there are any UI strings I should have a look at (why are the translation files updated?).
> 
> @gabrielmougard was this just missed in another PR ?
> 
> Are there actually any changes here?

no, there are no changes here

**Comment by tomponline** [comment]
> > > Still wondering if there are any UI strings I should have a look at (why are the translation files updated?).
> > 
> > 
> > @gabrielmougard was this just missed in another PR ?
> > Are there actually any changes here?
> 
> no, there are no changes here

OK no need for that commit then

**Review by ru-fu — APPROVED** [review]
Thanks!

**Pull request merged**
